executor/src/witgen/jit/compiler.rs

```diff
@@ -0,0 +1,539 @@
+    let (known_idx, known_bit) = index_known(global_offset, local_offset, column);
+    known[known_idx as usize] |= 1 << (known_bit);
+}}
+
+#[repr(C)]
+enum LookupCell<'a, T> {{
+    /// Value is known (i.e. an input)
+    Input(&'a T),
+    /// Value is not known (i.e. an output)
+    Output(&'a mut T),
+}}
+
+#[repr(C)]
+struct WitgenFunctionParams {{
+    data: *mut FieldElement,
+    known: *mut u32,
+    len: u64,
+    row_offset: u64,
+    call_machine: fn(u64, *mut [LookupCell<'_, FieldElement>], usize) -> bool,
+}}
+    "#
+    ))
+}
+
+#[cfg(test)]
+mod tests {
+    use pretty_assertions::assert_eq;
+
+    use powdr_number::GoldilocksField;
+
+    use super::*;
+
+    #[test]
+    fn compile_util_code() {
+        if let Err(e) = compile_effects::<GoldilocksField>(0, 2, &[], &[]) {
+            println!("{e}");
+            panic!()
+        }
+    }
+
+    fn cell(column_name: &str, id: u64, row_offset: i32) -> Cell {
+        Cell {
+            column_name: column_name.to_string(),
+            row_offset,
+            id,
+        }
+    }
+
+    fn symbol(cell: &Cell) -> SymbolicExpression<GoldilocksField, Cell> {
+        SymbolicExpression::from_symbol(cell.clone(), None)
+    }
+
+    fn number(n: u64) -> SymbolicExpression<GoldilocksField, Cell> {
+        SymbolicExpression::from(GoldilocksField::from(n))
+    }
+
+    fn assignment(
+        cell: &Cell,
+        e: SymbolicExpression<GoldilocksField, Cell>,
+    ) -> Effect<GoldilocksField, Cell> {
+        Effect::Assignment(cell.clone(), e)
+    }
+
+    #[test]
+    fn simple_effects() {
+        let a0 = cell("a", 2, 0);
+        let x0 = cell("x", 0, 0);
+        let ym1 = cell("y", 1, -1);
+        let yp1 = cell("y", 1, 1);
+        let effects = vec![
+            assignment(&x0, number(7) * symbol(&a0)),
+            assignment(&ym1, symbol(&x0)),
+            assignment(&yp1, symbol(&ym1) + symbol(&x0)),
+            Effect::Assertion(Assertion {
+                lhs: symbol(&ym1),
+                rhs: symbol(&x0),
+                expected_equal: true,
+            }),
+        ];
+        let known_inputs = vec![a0.clone()];
+        let code = witgen_code(&known_inputs, &effects);
+        assert_eq!(
+            code,
+            "
+#[no_mangle]
+extern \"C\" fn witgen(
+    WitgenFunctionParams{
+        data,
+        known,
+        len,
+        row_offset,
+        call_machine
+    }: WitgenFunctionParams,
+) {
+    let data = data_to_slice(data, len);
+    let known = known_to_slice(known, len);
+
+    let a_d0 = get(data, row_offset, 0, 2);
+
+    let x_d0 = (FieldElement::from(7) * a_d0);
+    let y_u1 = x_d0;
+    let y_d1 = (y_u1 + x_d0);
+    assert!(y_u1 == x_d0);
+
+    set(data, row_offset, 0, 0, x_d0);
+    set(data, row_offset, -1, 1, y_u1);
+    set(data, row_offset, 1, 1, y_d1);
+
+    set_known(known, row_offset, 0, 0);
+    set_known(known, row_offset, -1, 1);
+    set_known(known, row_offset, 1, 1);
```

Comment:
The main reason I did it like this is because I think this is the best form for the optimizer: First copy everything to local variables (i.e. the stack), operate only on stack variables and then copy to memory again. My assumption is that as long as stuff is on the stack, the optimizer knows that it cannot be modified. If data is in memory, any memory write can potentially destroy the data and we cannot re-use already computed values. This is particularly important for submachine calls: If values are on the stack, we can be sure that the are still the same after the call to the submachine. Also I think the optimizer is smart enough to directly store data in memory if it will not be touched again.

We should verify all these assumptions by looking at the assembly an the performance, of course.